external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/TokenUtil.scala

```diff
@@ -0,0 +1,116 @@
+import org.apache.kafka.common.security.token.delegation.DelegationToken
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+
+private[kafka010] object TokenUtil extends Logging {
+  val TOKEN_KIND = new Text("KAFKA_DELEGATION_TOKEN")
+  val TOKEN_SERVICE = new Text("kafka.server.delegation.token")
+
+  private class KafkaDelegationTokenIdentifier extends AbstractDelegationTokenIdentifier {
+    override def getKind: Text = TOKEN_KIND;
+  }
+
+  private def printToken(token: DelegationToken): Unit = {
+    val dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm")
+    log.info("%-15s %-30s %-15s %-25s %-15s %-15s %-15s".format(
+      "TOKENID", "HMAC", "OWNER", "RENEWERS", "ISSUEDATE", "EXPIRYDATE", "MAXDATE"))
+    val tokenInfo = token.tokenInfo
+    log.info("%-15s [hidden] %-15s %-25s %-15s %-15s %-15s".format(
+      tokenInfo.tokenId,
+      tokenInfo.owner,
+      tokenInfo.renewersAsString,
+      dateFormat.format(tokenInfo.issueTimestamp),
+      dateFormat.format(tokenInfo.expiryTimestamp),
+      dateFormat.format(tokenInfo.maxTimestamp)))
+  }
+
+  def obtainToken(sparkConf: SparkConf): Token[_ <: TokenIdentifier] = {
+    val adminClientProperties = new Properties
+
+    val bootstrapServers = sparkConf.get(KAFKA_BOOTSTRAP_SERVERS)
+    require(bootstrapServers.nonEmpty, s"Tried to obtain kafka delegation token but bootstrap " +
+      "servers not configured.")
+    adminClientProperties.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers.get)
+
+    val protocol = sparkConf.get(KAFKA_SECURITY_PROTOCOL)
+    adminClientProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, protocol)
+    if (protocol.endsWith("SSL")) {
+      logInfo("SSL protocol detected.")
+
+      val truststoreLocation = sparkConf.get(KAFKA_TRUSTSTORE_LOCATION)
+      if (truststoreLocation.nonEmpty) {
+        adminClientProperties.put("ssl.truststore.location", truststoreLocation.get)
+      } else {
+        logInfo("No truststore location set for SSL.")
+      }
+
+      val truststorePassword = sparkConf.get(KAFKA_TRUSTSTORE_PASSWORD)
+      if (truststorePassword.nonEmpty) {
+        adminClientProperties.put("ssl.truststore.password", truststorePassword.get)
+      } else {
+        logInfo("No truststore password set for SSL.")
+      }
+    } else {
+      logWarning("Obtaining kafka delegation token through plain communication channel. Please " +
+        "consider it's security impact.")
+    }
+
+    // There are multiple possibilities to log in:
+    // - Keytab is provided -> try to log in with kerberos module using kafka's dynamic JAAS
+    //   configuration.
+    // - Keytab not provided -> try to log in with JVM global security configuration
+    //   which can be configured for example with 'java.security.auth.login.config'.
+    //   For this no additional parameter needed.
+    KafkaSecurityHelper.getKeytabJaasParams(sparkConf) match {
```

Comment:
Same comment as before about not logging anything in the default case.

Reply:
Fixed.